### PR TITLE
perf: Improve xaml generator performance

### PR DESCRIPTION
**Pull request by Youssef1313**
Previously, every run of the generator was collecting a dictionary of all symbols in compilation references. This is very expensive, and the generator is run on every change in the IDE. This is both slow and very memory consuming which can in turn cause high GC pause times.

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/SolutionTemplate/Uno.ProjectTemplates.Dotnet/content/unoapp-uwp-net6/.vsconfig [hunk 1]
```diff
@@ -3,92 +3,36 @@
   "components": [
     "Microsoft.VisualStudio.Component.CoreEditor",
     "Microsoft.VisualStudio.Workload.CoreEditor",
-    "Microsoft.NetCore.Component.Runtime.3.1",
```
Comment (Youssef1313) [inline src/SolutionTemplate/Uno.ProjectTemplates.Dotnet/content/unoapp-uwp-net6/.vsconfig:6]: I didn't change anything in `.vsconfig`. These were done automatically when I built from VS.

**Comment by Youssef1313** [comment]
@jeromelaban This is ready for review.

**Comment by jeromelaban** [comment]
This is unfortunately a large breaking change. There are some legacy codebases which rely on the use of fuzzy matching.

This may be a change that we can include in the next major breaking change pass.

That being said, the change about the `Binding` attribute may be included sooner, as it is not breaking.

**Comment by Youssef1313** [comment]
I tried to reduce the impact of the breaking change by adding additional namespaces that are searched by default. We can also introduce an MSBuild property `UnoXamlDefaultNamespaces` to make it easier for developers to overcome the breaking change. So, if someone largely broke, they can add `<UnoXamlDefaultNamespace>CommaSeparatedNamespaces</UnoXamlDefaultNamespace>.

Alternatively, we can introduce a `UnoEnableXamlFuzzyMatching` property and document that it affects performance badly and that it will be removed in the next major release. This property will restore the fuzzy matching.

**Comment by Youssef1313** [comment]
Also worth noting that code relying on this wouldn't compile on Windows.

**Comment by Youssef1313** [comment]
Worth noting that the GC pause time I found was too large, and these allocations are the major contributor.

image

**Comment by jeromelaban** [comment]
I don't mind if we introduce this breaking change in the next major, it's a good one. Introducing the msbuild property is a good idea, we'll make it enabled by default for 4.x, and mention that in the docs that devs will need to test it when moving to the later version.

**Comment by Youssef1313** [comment]
> I don't mind if we introduce this breaking change in the next major, it's a good one. Introducing the msbuild property is a good idea, we'll make it enabled by default for 4.x, and mention that in the docs that devs will need to test it when moving to the later version.

Great! I'll add this property and default it to true.

**Comment by Youssef1313** [comment]
@jeromelaban Can you take another look?

**Comment by jeromelaban** [comment]
This flag also raises an additional blind spot where we suggest in the documentation to use namespaces to exclude platform specific code, but without providing a way to specify a namespace for lookups. We'll need to find a proper way to handle this, following what was discussed a while back in https://github.com/unoplatform/uno/issues/789

**Comment by Youssef1313** [comment]
> This flag also raises an additional blind spot where we suggest in the documentation to use namespaces to exclude platform specific code, but without providing a way to specify a namespace for lookups. We'll need to find a proper way to handle this, following what was discussed a while back in #789

I guess if one will use something from the default namespaces, then `xmlns:android="http://uno.ui/android"` should work. Otherwise, `xmlns:android="using:NamespaceHere"` will work (already supported I think? - but not documented)

And finally, if platform-specific code is needed with different namespaces, probably do one of the two things:

- Use `IsTypePresent` and pass android-only symbol.
- Use multiple namespaces, `<IncludeXamlNamespaces Include="android_MyNamespace" />` where in XAML this will be `xmlns:android_MyNamespace="using:MyNamespace`, etc for other namespaces.

@jeromelaban Does that sound good? If so, I can document that when removing the new flag in breaking changes branch.

**Review by agneszitte-nventive — APPROVED** [review]

**Comment by jeromelaban** [comment]
This is not enough, no, because `android` is meant to be just for one namespace, which means that developers will need to include more `android` like namespaces as defined here: 

https://github.com/unoplatform/uno/blob/8ddbd9e87c815d69403bc0136e8adbb4f3a2d84c/src/SourceGenerators/Uno.UI.SourceGenerators/Content/Uno.UI.SourceGenerators.props#L161-L190

This is not particular user-friendly, and we'll need to come up with another way to exclude namespaces based on a specific format, maybe in the Uri or something else, which was partially the discussion in https://github.com/unoplatform/uno/issues/789.

**Comment by Youssef1313** [comment]
Two different errors in CI. One that I cannot reproduce locally, and the other is because of incomplete `IncludeXamlNamespaces`/`ExcludeXamlNamespaces` in the outer build (not sure if that is the correct terminology, but I'm referring to the build that has all TFMs set but no specific TFM to be build, which at later point is dispatched to inner builds). The task runs on the build before it is dispatched so we don't (can't) have all xaml namespaces set. I'll update to rely on `Ignorable` as you suggested before @jeromelaban 

Then will see if the error that I cannot reproduce locally will happen.

**Comment by Youssef1313** [comment]
@jeromelaban Finally, this is green now 🎉

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/ListView/HorizontalListView_Templates.xaml [hunk 2]
```diff
@@ -9,7 +9,7 @@
 	xmlns:ios="http://uno.ui/ios"
 	xmlns:uConverters="using:Uno.UI.Samples.Converters"
 	xmlns:win="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
-	xmlns:android="http://uno.ui/android"
+	xmlns:android="http://uno.ui/android#using:Uno.UI.Controls.Legacy"
```
Comment (jeromelaban) [inline src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/ListView/HorizontalListView_Templates.xaml:12]: We need to document this in https://github.com/unoplatform/uno/blob/master/doc/articles/platform-specific-xaml.md

File: src/Directory.Build.props [hunk 3]
```diff
@@ -58,6 +58,8 @@
 
     <DisableImplicitNamespaceImports>true</DisableImplicitNamespaceImports>
 
+    <UnoEnableXamlFuzzyMatching>true</UnoEnableXamlFuzzyMatching>
```
Comment (Youssef1313) [inline src/Directory.Build.props:61]:
```suggestion
    <UnoEnableXamlFuzzyMatching>false</UnoEnableXamlFuzzyMatching>
```

File: src/SourceGenerators/Uno.UI.SourceGenerators/BindableTypeProviders/BindableTypeProvidersGenerationTask.cs [hunk 4]
```diff
@@ -85,9 +84,8 @@ internal void Generate(GeneratorExecutionContext context)
 						_assemblyName = context.GetMSBuildPropertyValue("AssemblyName");
 
 						_defaultNamespace = context.GetMSBuildPropertyValue("RootNamespace");
-						_namedSymbolsLookup = context.Compilation.GetSymbolNameLookup();
 
-						_bindableAttributeSymbol = FindBindableAttributes();
+						_bindableAttributeSymbol = context.Compilation.GetTypeByMetadataName(XamlConstants.Types.BindableAttribute);
```
Comment (jeromelaban) [inline src/SourceGenerators/Uno.UI.SourceGenerators/BindableTypeProviders/BindableTypeProvidersGenerationTask.cs:88]: We should also be searching for `System.ComponentModel.BindableAttribute`, as a compatibility measure as there are some non-winui viewmodels that use it.

**Pull request merged**